File: 4/main.cpp
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>
#include<WinSock.h>
#pragma comment(lib, "Ws2_32.lib")

void MAC_print(FILE *out, unsigned char *MAC);
void IP_print(FILE *out, unsigned char *IP);

int main() {
    FILE *in = NULL;
    FILE *out = NULL;
    bool enterfile = false;
    char filename[100];
    int filesize = 0;
    int framenumber = 1;
    int frame_ip = 0, frame_arp = 0, frame_snap = 0, frame_8023 = 0;
    int dix = 0;
    unsigned char *DATA;

    // Ввод имени файла
    while(!enterfile) {
        printf("File name: ");
        fgets(filename, sizeof(filename), stdin);
        filename[strcspn(filename, "\n")] = 0;  // Убираем символ новой строки
        in = fopen(filename, "rb");
        if(in != NULL) enterfile = true;
        else printf("This file does not exist! Try again.\n\n");
    }

    // Открытие файла для вывода
    out = fopen("out.txt", "w");

    // Чтение данных из файла
    fseek(in, 0, SEEK_END);
    filesize = ftell(in);
    fseek(in, 0, SEEK_SET);
    DATA = (unsigned char*)malloc(filesize);
    fread(DATA, filesize, 1, in);
    fclose(in);

    fprintf(out, "Size of file: %d bytes\n", filesize);
    fprintf(out, "\n----------------\n");

    unsigned char *p = DATA;

    while(p < DATA + filesize) {
        fprintf(out, "Frame number: %d\n", framenumber);
        fprintf(out, "MAC dest: ");
        MAC_print(out, p);  // MAC адрес назначения
        fprintf(out, "MAC source: ");
        MAC_print(out, p + 6);  // MAC адрес источника
        
        // Проверка поля длины/типа (13-й и 14-й байты)
        unsigned short LT = ntohs(*(unsigned short*)(p + 12));  // Правильное чтение 16-битного значения
        
        // Если значение поля типа протокола больше 0x05DC, это Ethernet II
        if (LT == 0x0800) {  // IPv4
            fprintf(out, "Frame type: IPv4\n");
            fprintf(out, "IP source: ");
            IP_print(out, p + 26);  // IP адрес источника
            fprintf(out, "IP dest: ");
            IP_print(out, p + 30);  // IP адрес назначения
            frame_ip++;
            p += ntohs(*(unsigned short*)(p + 16)) + 14;  // Длина кадра IPv4 + заголовок
        } else if (LT == 0x0806) {  // ARP
            fprintf(out, "Frame type: ARP\n");
            frame_arp++;
            p += 42;  // Длина ARP кадра
        } else if (LT >= 0x8137 && LT <= 0x8138) {  // Novell IPX
            fprintf(out, "Frame type: Novell IPX\n");
            p += 30;
        } else if (LT == 0x8100) {  // Ethernet VLAN
            fprintf(out, "Frame type: Ethernet VLAN\n");
            p += 18;  // Длина VLAN кадра
        } else if (LT > 0x05FE) {  // Если значение больше порога, это Ethernet II
            fprintf(out, "Frame type: Ethernet DIX (Ethernet II)\n");
            dix++;
        } else {  
            // Этап 3: Если значение меньше или равно 0x05DC, продолжаем проверку
            unsigned short F = ntohs(*(unsigned short*)(p + 14));
            
            // Этап 4: Если первые два байта равны 0xFFFF, это Ethernet_802.3 для NetWare 3.x
            if (F == 0xFFFF) {
                fprintf(out, "Frame type: Raw 802.3 (Ethernet 802.3)\n");
                frame_8023++;
            }
            // Этап 5: Если первые два байта равны 0xAAAA, это Ethernet_SNAP
            else if (F == 0xAAAA) {
                fprintf(out, "Frame type: Ethernet SNAP\n");
                frame_snap++;
            }
            // Этап 6: Иначе, это стандартный Ethernet 802.2 (LLC)
            else {
                fprintf(out, "Frame type: 802.3/LLC (Ethernet 802.2)\n");
                frame_8023++;
            }
        }

        fprintf(out, "\n---------\n");
        p += 60;  // Пропускаем стандартный кадр (если кадры разной длины, этот участок должен быть изменен)
        framenumber++;
    }

    fprintf(out, "Total number of frames: %d\n", framenumber - 1);
    fprintf(out, "IPv4 frames: %d\n", frame_ip);
    fprintf(out, "Dix frames: %d\n", dix);
    fprintf(out, "ARP frames: %d\n", frame_arp);
    fprintf(out, "SNAP frames: %d\n", frame_snap);
    fprintf(out, "802.3 frames: %d\n", frame_8023);

    fclose(out);
    free(DATA);
}

void MAC_print(FILE *out, unsigned char *MAC) {
    for (int i = 0; i < 6; i++) {
        fprintf(out, "%02X", MAC[i]);
        if (i < 5) fprintf(out, ":");
    }
    fprintf(out, "\n");
}

void IP_print(FILE *out, unsigned char *IP) {
    for (int i = 0; i < 4; i++) {
        fprintf(out, "%d", IP[i]);
        if (i < 3) fprintf(out, ".");
    }
    fprintf(out, "\n");
}

File: 4/main.py
# Определяем протоколы
PROT_IPV4 = 0x0800
PROT_ARP = 0x0806
PROT_IPV6 = 0x08DD
PROT_IPX = 0x8137

input_file = input("Введите название файла для чтения: ")
try:
    with open(input_file, "rb") as f:
        buffer = bytearray(f.read()) # Читаем весь файл в буфер
        print(f"Файл \"{input_file}\" успешно считан.")
except FileNotFoundError:
    print(f"Файл \"{input_file}\" отсутствует.")
    exit(0)

count_of_frames = 0  # Счетчик фреймов
count_of_frames_type = [0] * 4  # Кол-во обработанных типов фреймов
count_of_protocol = [0] * 3  # Кол-во обработанных типов протоколов

print(f"\n{'-' * 30}\n")
current = 0  # Текущий элемент
total_size = 0  # Общий размер

while current < len(buffer):
    count_of_frames += 1
    print(f"Фрейм {count_of_frames}")

    print("MAC-адрес получателя: ")
    print(':'.join(f"{buffer[current + i]:02X}" for i in range(6)))

    print("MAC-адрес отправителя: ")
    print(':'.join(f"{buffer[current + 6 + i]:02X}" for i in range(6)))

    # Чтение типа протокола или длины
    len_type = int(buffer[current + 12:current + 14].hex(), 16)
    if len_type > 0x05DC:
        count_of_frames_type[0] += 1
        print("Тип фрейма: Ethernet II")

        if len_type == PROT_IPV4:
            count_of_protocol[0] += 1
            version = (buffer[current + 14] & 0xF0) >> 4
            head_size = (buffer[current + 14] & 0x0F) * 4
            body_size = int(buffer[current + 16:current + 18].hex(), 16)

            print(f"Версия: {version}")
            print("IP-адрес источника: ")
            print('.'.join(str(buffer[current + 26 + i]) for i in range(4)))

            print("IP-адрес назначения: ")
            print('.'.join(str(buffer[current + 30 + i]) for i in range(4)))

            len_type = body_size + 14  # Учитываем заголовок
            print(f"Размер фрейма: {len_type} байт")

            protocol = buffer[current + 23] # определяем протокол
            if protocol == 6:
                print("Протокол: TCP")
            elif protocol == 17:
                print("Протокол: UDP")

            print("Данные: ")
            for i in range(body_size - head_size):
                print(f"{buffer[current + 14 + head_size + i]:02X} ", end='')
            print()
            current += len_type
            total_size += len_type

        elif len_type == PROT_ARP:
            count_of_protocol[1] += 1
            print("Тип фрейма: ARP")

            print("MAC-адрес отправителя: ")
            print(':'.join(f"{buffer[current + 22 + i]:02X}" for i in range(6)))

            print("IP-адрес отправителя: ")
            print('.'.join(str(buffer[current + 28 + i]) for i in range(4)))

            print("MAC-адрес получателя: ")
            print(':'.join(f"{buffer[current + 32 + i]:02X}" for i in range(6)))

            print("IP-адрес получателя: ")
            print('.'.join(str(buffer[current + 38 + i]) for i in range(4)))

            print("Размер фрейма: 42 байт")
            current += 42
            total_size += 42

        elif len_type == PROT_IPV6:
            print("Тип протокола: IPv6")
        elif len_type == PROT_IPX:
            print("Тип протокола: IPX (Novell)")
        else:
            print("Тип протокола: Неизвестен.")
            print(f"Номер протокола: {len_type:x}")
            current += len_type + 14
            total_size += len_type + 14
            count_of_protocol[2] += 1
    else:
        count_of_protocol[2] += 1
        llc = (buffer[current + 14] << 8) + buffer[current + 15] # тип протокола LLC
        if llc == 0xFFFF:
            count_of_frames_type[1] += 1
            print("Тип фрейма: Ethernet_802.3 для NetWare 3.х")
            print(f"Размер фрейма: {len_type + 14} байт")
            print("Данные: ")
            for i in range(len_type):
                print(f"{buffer[current + 14 + i]:02X} ", end='')
            print()
        elif llc == 0xAAAA:
            count_of_frames_type[2] += 1
            print("Тип фрейма: Ethernet_SNAP")
            print(f"Размер фрейма: {len_type + 14} байт")
            print("Данные: ")
            for i in range(len_type - 3 - 5):
                print(f"{buffer[current + 14 + 3 + 5 + i]:02X} ", end='')
            print()
        else:
            count_of_frames_type[3] += 1
            print("Тип фрейма: Ethernet_802.2")
            print(f"Размер фрейма: {len_type + 14} байт")
            print("Данные: ")
            for i in range(len_type - 3):
                print(f"{buffer[current + 14 + 3 + i]:02X} ", end='')
            print()
        current += len_type + 14
        total_size += len_type + 14

    print(f"\n{'-' * 30}\n")

#Результаты
print(f"Считано: {count_of_frames} фреймов")
print("Типы фреймов:")
print(f"Ethernet II: {count_of_frames_type[0]}")
print(f"Ethernet_802.3 для NetWare 3.х: {count_of_frames_type[1]}")
print(f"Ethernet_SNAP: {count_of_frames_type[2]}")
print(f"Ethernet_802.2: {count_of_frames_type[3]}")
print("Типы протоколов:")
print(f"IPv4: {count_of_protocol[0]}")
print(f"ARP: {count_of_protocol[1]}")
print(f"Неопределен: {count_of_protocol[2]}")
